api/http-api.md

```diff
@@ -0,0 +1,176 @@
+  * Do not assume any names reserved. So we can't have `/workflows/{workflow_id}` and `/workflows/general-call` in case
+    there's a workflow named `general-call`.
+  * Use plural nouns where an identifier may be provided to differentiate
+    * But singular nouns are not _required_ in other places, can still be plural
+* Response codes
+  * 201 success if a creation
+  * 202 if action eventually consistent
+  * 200 other all other successes
+  * Do not use 204 for no response, even if there's no response because we may add response later, just 200 everything
+* Methods:
+  * `GET` if read-only
+    * Noun as last URL term
+    * Use query parameters for commonly set items to avoid relying on request bodies (but still rely on bodies as
+      needed). This may require some gRPC gateway customization.
+  * `PUT` if strongly consistent add/update on fixed-URL w/ idempotency
+    * Noun as last URL term
+  * `DELETE` if item is actually deleted (i.e. unreachable from `GET`)
+    * Noun as last URL term
+  * `POST` for everything else
+    * Noun as last URL term if for a create, verb if for an action
+  * `PATCH` is not needed because we don't adhere to common CRUD
+    * TODO(cretz): Needs discussion
+
+### RPCs
+
+Below is list of gRPC calls to HTTP calls so far and some notes/questions beneath each. Some notes/questions beneath one
+actually apply to many but we are not repeating ourselves too much here.
+
+* Namespaces:
+  * `RegisterNamespace` - `POST /api/v1/namespaces`
+    * 202 on success since eventually consistent
+  * `DescribeNamespace` - `GET /api/v1/namespaces/{namespace}`
+  * `ListNamespaces` - `GET /api/v1/namespaces`
+  * `UpdateNamespace` - `PUT /api/v1/namespaces/{namespace}`
+    * TODO(cretz): Confirm strongly consistent and idempotent or not
+  * `DeprecateNamespace` - `DELETE /api/v1/namespaces/{namespace}`
+    * Is this eventually consistent? Will determine response code
+    * Does this prevent returning from `GET`? Will determine HTTP method
+* Workflows:
+  * `StartWorkflowExecution` - `POST /api/v1/namespaces/{namespace}/workflows/{workflow_id}`
+    * 204 on success
+    * Should accept header as `identity` on these calls?
+    * Ok to be at root URL here instead of appending `/start`?
+  * `GetWorkflowExecutionHistory` - `GET /api/v1/namespaces/{namespace}/workflows/{workflow_id}/history`
+    * Accept `?nextPageToken` as query param? I'm thinking so
+  * `GetWorkflowExecutionHistoryReverse` - `GET /api/v1/namespaces/{namespace}/workflows/{workflow_id}/history-reverse`
+  * `RequestCancelWorkflowExecutionRequest` - `POST /api/v1/namespaces/{namespace}/workflows/{workflow_id}/cancel`
+    * Intentionally not taking the run ID in the URL, but should it be `?runId=` query param?
+  * `SignalWorkflowExecution` - `POST /api/v1/namespaces/{namespace}/workflows/{workflow_id}/signal`
+    * Should we put signal name in the URL? Can be optional (e.g. in URL or body)
```

Comment:
What if the entire signal could be sent within either the URL or body? It makes linking to interact with human-in-the-loop workflows nice and simple. But also allows for more complex signals to be sent in the body of the POST request.

I was going to post a similar question for async activity completion but will keep the conversation here since it is a similar purpose.

Reply:
> What if the entire signal could be sent within either the URL or body?

Yes, this was the subject of the question and the parenthetical. I am thinking optional signal name in URL makes sense. Note that doesn't help signal args which I think goes a step too far. This can apply to queries too. But sadly not updates that have their own stable identifier.

Also note this is a POST and a thin layer over gRPC. Simplicity for humans is not as large of a factor (part of the non-goal of designing a better API).

> I was going to post a similar question for async activity completion but will keep the conversation here since it is a similar purpose.

This is a different deal. To refer to an activity you need a task token or an ID. The former is what we recommend but is too large and unwieldy to put on a URL and therefore we don't want to make the latter easier. Also, we discourage async activities unless they are really needed, so we're not trying to design easy HTTP API approaches for their use beyond surfacing the gRPC API.

Reply:
Updated to allow optional signal name on URL. Made optional because names can get unwieldy and not URL friendly, but I'm starting to consider just making them required.

Reply:
Resolving...